File: app/src/main/java/com/arturogutierrez/openticator/domain/navigator/drawer/NavigationDrawer.java
package com.arturogutierrez.openticator.domain.navigator.drawer;

import android.app.Activity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.arturogutierrez.openticator.R;
import com.arturogutierrez.openticator.domain.category.model.Category;
import com.mikepenz.materialdrawer.Drawer;
import com.mikepenz.materialdrawer.DrawerBuilder;
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem;
import com.mikepenz.materialdrawer.model.SectionDrawerItem;
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem;
import java.util.List;
import javax.inject.Inject;

public class NavigationDrawer implements NavigationDrawerView {

  @Bind(R.id.toolbar)
  Toolbar toolbar;

  private final Activity activity;
  private final NavigationDrawerPresenter presenter;

  private Drawer drawer;
  private SectionDrawerItem categoriesSectionDrawerItem;
  private PrimaryDrawerItem allAccountsDrawerItem;

  @Inject
  public NavigationDrawer(Activity activity, NavigationDrawerPresenter presenter) {
    this.activity = activity;
    this.presenter = presenter;

    initialize(activity);
  }

  public void onCreate(Bundle savedInstanceState) {
    createDrawerLayout(savedInstanceState);
  }

  public void onResume() {
    presenter.resume();
  }

  public void onPause() {
    presenter.pause();
  }

  public void onDestroy() {
    presenter.destroy();
  }

  public boolean onBackPressed() {
    if (drawer != null && drawer.isDrawerOpen()) {
      drawer.closeDrawer();
      return true;
    }

    return false;
  }

  private void initialize(Activity activity) {
    ButterKnife.bind(this, activity);
    presenter.setView(this);
  }

  private void createDrawerLayout(Bundle savedInstanceState) {
    allAccountsDrawerItem = new PrimaryDrawerItem().withName("All accounts")
        .withIcon(R.drawable.ic_folder_black_24dp)
        .withIconTintingEnabled(true);
    categoriesSectionDrawerItem = new SectionDrawerItem().withName("Categories").withDivider(false);

    drawer = new DrawerBuilder().withSavedInstance(savedInstanceState)
        .withActivity(activity)
        .withToolbar(toolbar)
        .withActionBarDrawerToggleAnimated(true)
        .addDrawerItems(categoriesSectionDrawerItem, allAccountsDrawerItem)
        .withOnDrawerItemClickListener(
            (view, position, drawerItem) -> onDrawerItemClicked(drawerItem, position))
        .withSelectedItemByPosition(1)
        .build();
  }

  @Override
  public void renderCategories(List<Category> categories) {
    drawer.removeAllItems();
    drawer.addItems(categoriesSectionDrawerItem, allAccountsDrawerItem);

    for (Category category : categories) {
      PrimaryDrawerItem drawerItem = createCategoryDrawerItem(category);
      drawer.addItem(drawerItem);
    }
  }

  @Override
  public void dismissDrawer() {
    drawer.closeDrawer();
  }

  private PrimaryDrawerItem createCategoryDrawerItem(Category category) {
    return new PrimaryDrawerItem().withName(category.getName()).withLevel(5);
  }

  private boolean onDrawerItemClicked(IDrawerItem drawerItem, int position) {
    if (drawerItem == allAccountsDrawerItem) {
      presenter.allCategoriesItemClicked();
    } else {
      presenter.categoryItemClicked(position - 2);
    }
    return true;
  }
}

File: app/src/main/kotlin/com/arturogutierrez/openticator/domain/navigator/drawer/NavigationDrawer.kt
package com.arturogutierrez.openticator.domain.navigator.drawer

import android.app.Activity
import android.os.Bundle
import android.support.v7.widget.Toolbar
import com.arturogutierrez.openticator.R
import com.arturogutierrez.openticator.domain.category.model.Category
import com.arturogutierrez.openticator.domain.navigator.Navigator
import com.arturogutierrez.openticator.helpers.find
import com.mikepenz.materialdrawer.Drawer
import com.mikepenz.materialdrawer.DrawerBuilder
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem
import com.mikepenz.materialdrawer.model.SectionDrawerItem
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem
import javax.inject.Inject

class NavigationDrawer @Inject constructor(val activity: Activity,
    val presenter: NavigationDrawerPresenter,
    val navigator: Navigator) : NavigationDrawerView {

  private val toolbar by lazy { activity.find<Toolbar>(R.id.toolbar) }
  private val allAccountsDrawerItem = createAllAccountsDrawerItem()
  private val categoriesSectionDrawerItem = createCategoriesSectionDrawerItem()
  private val settingsDrawerItem = createSettingsDrawerItem()

  private lateinit var drawer: Drawer

  fun onCreate(savedInstanceState: Bundle?) {
    createDrawerLayout(savedInstanceState)
  }

  fun onResume() {
    presenter.attachView(this)
    presenter.resume()
  }

  fun onPause() {
    presenter.detachView()
    presenter.pause()
  }

  fun onDestroy() {
    presenter.destroy()
  }

  fun onBackPressed(): Boolean {
    if (drawer.isDrawerOpen) {
      drawer.closeDrawer()
      return true
    }

    return false
  }

  private fun createDrawerLayout(savedInstanceState: Bundle?) {
    drawer = DrawerBuilder()
        .withSavedInstance(savedInstanceState)
        .withActivity(activity)
        .withToolbar(toolbar)
        .withActionBarDrawerToggleAnimated(true)
        .addDrawerItems(categoriesSectionDrawerItem, allAccountsDrawerItem)
        .addStickyDrawerItems(settingsDrawerItem)
        .withOnDrawerItemClickListener { _, position, drawerItem -> onDrawerItemClicked(drawerItem, position) }
        .withSelectedItemByPosition(1)
        .build()
  }

  override fun renderCategories(categories: List<Category>) {
    drawer.removeAllItems()
    drawer.addItems(categoriesSectionDrawerItem, allAccountsDrawerItem)

    categories.map { createCategoryDrawerItem(it) }
        .forEach { drawer.addItem(it) }
  }

  override fun dismissDrawer() {
    drawer.closeDrawer()
  }

  private fun createSettingsDrawerItem(): PrimaryDrawerItem {
    return PrimaryDrawerItem()
        .withName(R.string.settings)
        .withIcon(R.drawable.ic_settings_black_24dp)
        .withIconTintingEnabled(true)
  }

  private fun createAllAccountsDrawerItem(): PrimaryDrawerItem {
    return PrimaryDrawerItem()
        .withName(R.string.all_accounts)
        .withIcon(R.drawable.ic_folder_black_24dp)
        .withIconTintingEnabled(true)
  }

  private fun createCategoriesSectionDrawerItem(): SectionDrawerItem {
    return SectionDrawerItem()
        .withName(R.string.categories)
        .withDivider(false)
  }

  private fun createCategoryDrawerItem(category: Category): PrimaryDrawerItem {
    return PrimaryDrawerItem().withName(category.name).withLevel(5)
  }

  private fun onDrawerItemClicked(drawerItem: IDrawerItem<*, *>, position: Int): Boolean {
    when (drawerItem) {
      allAccountsDrawerItem -> presenter.allCategoriesItemClicked()
      settingsDrawerItem -> navigator.goToSettings(activity)
      else -> presenter.categoryItemClicked(position - 2)
    }
    return true
  }
}
